source/extensions/filters/network/kafka/serialization.h

```diff
@@ -0,0 +1,772 @@
+      required_ = length_buf_.get();
+
+      if (required_ == NULL_BYTES_LENGTH) {
+        ready_ = true;
+      }
+      if (required_ < NULL_BYTES_LENGTH) {
+        throw EnvoyException(fmt::format("invalid NULLABLE_BYTES length: {}", required_));
+      }
+
+      length_consumed_ = true;
+    }
+
+    if (ready_) {
+      return length_consumed;
+    }
+
+    const size_t data_consumed = std::min<size_t>(required_, remaining);
+    required_ -= data_consumed;
+
+    buffer += data_consumed;
+    remaining -= data_consumed;
+
+    if (required_ == 0) {
+      ready_ = true;
+    }
+
+    return length_consumed + data_consumed;
+  }
+
+  bool ready() const { return ready_; }
+
+  /**
+   * Returns length of ignored array, or -1 if that was null
+   */
+  INT32 get() const { return required_max_; }
+
+private:
+  constexpr static INT32 NULL_BYTES_LENGTH{-1};
+
+  Int32Buffer length_buf_;
+  bool length_consumed_{false};
+  INT32 required_max_;
+  INT32 required_;
+  bool ready_{false};
+};
+
+/**
+ * This buffer captures the data fed
+ */
+class NullableBytesCapturingBuffer : public Deserializer<NULLABLE_BYTES> {
+public:
+  size_t feed(const char*& buffer, uint64_t& remaining) {
+    const size_t length_consumed = length_buf_.feed(buffer, remaining);
+    if (!length_buf_.ready()) {
+      // break early: we still need to fill in length buffer
+      return length_consumed;
+    }
+
+    if (!length_consumed_) {
+      required_ = length_buf_.get();
+
+      if (required_ >= 0) {
+        data_buf_ = std::vector<unsigned char>(required_);
+      }
+      if (required_ == NULL_BYTES_LENGTH) {
+        ready_ = true;
+      }
+      if (required_ < NULL_BYTES_LENGTH) {
+        throw EnvoyException(fmt::format("invalid NULLABLE_BYTES length: {}", required_));
+      }
+
+      length_consumed_ = true;
+    }
+
+    if (ready_) {
+      return length_consumed;
+    }
+
+    const size_t data_consumed = std::min<size_t>(required_, remaining);
+    const size_t written = data_buf_.size() - required_;
+    memcpy(data_buf_.data() + written, buffer, data_consumed);
+    required_ -= data_consumed;
+
+    buffer += data_consumed;
+    remaining -= data_consumed;
+
+    if (required_ == 0) {
+      ready_ = true;
+    }
+
+    return length_consumed + data_consumed;
+  }
+
+  bool ready() const { return ready_; }
+
+  NULLABLE_BYTES get() const {
+    if (NULL_BYTES_LENGTH == required_) {
+      return absl::nullopt;
+    } else {
+      return {data_buf_};
+    }
+  }
+
+private:
+  constexpr static INT32 NULL_BYTES_LENGTH{-1};
+
+  Int32Buffer length_buf_;
+  bool length_consumed_{false};
+  INT32 required_;
+
+  std::vector<unsigned char> data_buf_;
+  bool ready_{false};
+};
+
+// === COMPOSITE BUFFER ========================================================
+
+/**
+ * Composes several buffers into one.
+ * The returned value is constructed via { buffer1.get(), buffer2.get() ... }
+ */
+template <typename RT, typename...> class CompositeBuffer;
```

Comment:
I just learned I can specialize this like
```
template <class... Ts> struct tuple {};

template <class T, class... Ts>
struct tuple<T, Ts...> : tuple<Ts...> {
  tuple(T t, Ts... ts) : tuple<Ts...>(ts...), tail(t) {}
  T tail;
};
```

Reply:
The most-offending `Composite` deserializer template has been removed and replaced with explicit domain deserializers.
All other deserializers are subclasses of `Deserializer<T>` interface and deserialize into instance of T.

The encoding template functions have been documented - in general I wanted to have single method that could work as `encode(string)`, `encode(int8)`, `encode(bool)` vs having `encodeString`/`encodeInt8`/`encodeBool` as IMHO it makes it harder to make generic code

Reply:
Sounds good. Let's see how it looks on the next pass.